### PR TITLE
Add function-url-scheme-whitelist

### PR DESCRIPTION
**Pull request by evilebottnawi**
/cc @davidtheclark @jeddy3

### CODE REVIEW

**Comment by jeddy3** [comment]
Umm. Is there an issue for this feature?

I'm not sure what everyone else thinks, but it might be a good idea if we focus on [bugs](https://github.com/stylelint/stylelint/issues?q=is%3Aissue+is%3Aopen+label%3A%22type%3A+bug%22), [the `7.0.0` milestone](https://github.com/stylelint/stylelint/milestones/7.0.0) and only new features associated with the [`7.0.0` release planning](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/release-planning.md) as they'll help with a smoother transition to `7.0.0`?

**Comment by evilebottnawi** [comment]
@jeddy3 i known, tomorrow i concentrate on issue v7, this rules we were very necessary for our team, sorry :smile:

**Comment by davidtheclark** [comment]
@evilebottnawi Can you please link to the issue where we discussed this at length? I remember there was a lot of discussion and would like to re-read it.

**Comment by ntwb** [comment]
@davidtheclark The discussion was in #1084

**Comment by davidtheclark** [comment]
Thanks @ntwb. Looks like my concern there hasn't really been addressed yet:

> The only one I'm uncertain about is `function-url-path: "absolute"|"relative"`. I'm having some difficulty envisioning the circumstance where you can/should only allow one or the other.

@evilebottnawi what do you say?

**Comment by evilebottnawi** [comment]
@davidtheclark disable absolute link, only relative allow, short url -> less bytes,

**Comment by davidtheclark** [comment]
As discussed in that issue, though, this would enforce never using `https`, which is a bad idea.

File: src/rules/function-url-path/__tests__/index.js [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    code: "a { background: url(./file.jpg); }",
+  }, {
+    code: "a { background: url(../file.jpg); }",
+  }, {
+    code: "a { background: URL(../file.jpg); }",
+  }, {
+    code: "a { background: url('../file.jpg'); }",
+  }, {
+    code: "a { background: url(\"../file.jpg\"); }",
+  }, {
```
Comment (ntwb) [inline src/rules/function-url-path/__tests__/index.js]: I just added a `https` test to `relative`:  `code: "a { background: url('https://www.google.com/file.jpg'); }",`

This test fails

File: src/rules/function-url-path/__tests__/index.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    message: messages.rejected,
+    line: 1,
+    column: 5,
+  }, {
+    code: "a { background: url(\"/file.jpg\"); }",
+    message: messages.rejected,
+    line: 1,
+    column: 5,
+  }, {
+    code: "a { background: url('//www.google.com/file.jpg'); }",
```
Comment (ntwb) [inline src/rules/function-url-path/__tests__/index.js]: This is an _anti-pattern_ and should be avoided, see https://www.paulirish.com/2010/the-protocol-relative-url/

Comment (evilebottnawi) [inline src/rules/function-url-path/__tests__/index.js]: @ntwb i agree this anti-pattern, but i think for test it is good

Comment (evilebottnawi) [inline src/rules/function-url-path/__tests__/index.js]: @ntwb for disable use this anti-pattern we have rule https://github.com/stylelint/stylelint/pull/1576

**Comment by ntwb** [comment]
Via https://github.com/stylelint/stylelint/issues/1084#issuecomment-212924186
> _"My understanding is that we don't want a `relative` option, because it's been deemed a bad practice. So we shouldn't build a rule that enforces it."_

This is a valid function-url-path: `a { background: url('https://www.google.com/file.jpg'); }`

Whether the option is `relative` or `absolute` does not matter, it **is a valid** URL path.

Per the comment I quoted above I believe this case still stands that we should *not* be enforcing `relative` URLs

____

I think part of the confusion here is the terminology `relative`, what is being implemented and tested here currently in this pull request is _both_ "URLs" **and** "File URLs" and the context of `relative` in relation to each of these is being interchanged causing confusion.

What I propose is to remove *all* "URL" context from this rule and only include "File URLs" and rename the rule to `function-url-file-path`:

* A valid `absolute` `a { background: url('/path/to/file.jpg'); }`
* A valid `relative`: `a { background: url(./file.jpg); }`

The above made sense after seeing it documented here: https://url.spec.whatwg.org/#syntax-url-relative

**Comment by davidtheclark** [comment]
Can't you still request files over https from a remote source? And how would we know if it's a "file URL" or not?

**Comment by ntwb** [comment]
> Can't you still request files over https from a remote source? And how would we know if it's a "file URL" or not?

Yes, that's what I get for writing that at 2am 😞 

So, ignore the 2nd part of my my comment....

The first part is still valid, I suggest we close this pull request and issue.

**Comment by jeddy3** [comment]
So, lets take a step back with this one.

What are we trying to achieve overall with our `function-url-*` rules?  

So far we can disallow [scheme-relative URLs](https://url.spec.whatwg.org/#syntax-url-scheme-relative) with `function-url-no-scheme-relative: true`.

The next step could be to specify *which* (*if any*) schemes are allowed:

- `function-url-scheme-blacklist: []`
- `function-url-scheme-whitelist: []`

Valid with ``function-url-scheme-whitelist: ["https", "file]`:

```
https://example.com/path/to/file.jpg // whitelisted
file:///path/to/file.jpg // whitelisted
./path/file.jpg // scheme-less, so not checked
/path/to/file.jpg // scheme-less, so not checked
```

You can disallow *all* schemes using an empty whitelist (like we do with other whitelist rules like `unit-whitelist`). If you choose to do this, you'd probably want to disallow either path-relative or path-absolute:

- `function-url-no-path-relative:  true`
- `function-url-no-path-absolute:  true`

Valid with `function-url-no-path-relative:  true`:

```
/path/to/file.jpg
```

Valid with `function-url-no-path-absolute:  true`:

```
./path/to/file.jpg
```

So, use cases?

Only want to allow https URLs (e.g. `http://example.com/file.jpg`):

```
function-url-scheme-whitelist: ["https],
function-url-no-path-absolute:  true,
function-url-no-path-relative:  true,
function-url-no-scheme-relative: true,
```

Only want to allow path-relative urls (e.g. `./path/to/file.jpg`)

```
function-url-scheme-whitelist: [],
function-url-no-path-absolute:  true,
function-url-no-scheme-relative: true,
```

What to allow https URLs and path-absolute ones (e.g. `http://example.com/file` & `/path/to/file.jpg`):

```
function-url-scheme-whitelist: ["https],
function-url-no-path-relative:  true,
function-url-no-scheme-relative: true,
```

Can anyone else think of any use cases not covered by the addition of these 4 new rules to `function-url-no-scheme-relative`?

P.S. I hope I've got the terminology right?... I've been trying to get my head around the URL spec for a while now!

**Comment by davidtheclark** [comment]
First: I think 5 rules is way too much energy dedicated to this. I'm pretty sure that people should  be using the URL schema most appropriate to the resource, which we can't know, instead of trying to make a blanket regulation.

For example, I'm pretty sure that `function-url-scheme-whitelist` should *only* accept `https`, right? Because it's always best practice to use that if it's available, and you should always try to make it available if it is not, etc.

**Comment by jeddy3** [comment]
>  I'm pretty sure that people should be using the URL schema most appropriate to the resource, *which we can't know*, instead of trying to make a blanket regulation.

That's a good point.

I think the origin of all this `function-url-*` stuff is the [UrlFormat](https://github.com/brigade/scss-lint/blob/master/lib/scss_lint/linter/README.md#urlformat) scss-lint rule, which the 18F guys made [a stylelint plugin](https://github.com/18F/stylelint-rules/blob/master/src/plugins/url-format/index.js) equivalent of. But, as you said, we can't know what scheme is the most appropriate to the resources e.g. loading a font from Google Fonts then you'll have to use a "domain" and "protocol". 

So, shall we close this PR and any further `function-url-*` things can be handled as plugins?

**Comment by davidtheclark** [comment]
I'm not sure about this myself. I guess if people are making plugins to accomplish this, they feel like it works. I guess it doesn't hurt (beyond PR review time) to accept these rules if someone wants to build them.

**Comment by jeddy3** [comment]
> I guess it doesn't hurt (beyond PR review time) to accept these rules if someone wants to build them.

Seems reasonable. Shall we just start with the `function-url-scheme-whitelist` rule as that, when given an empty array, achieves the same as the`url-format` plugin?

I won't create issues for the other rules until there's a request for them e.g. "How do I disallow path-absolute URLs (`/path/to/file.jpg`)?"

**Comment by davidtheclark** [comment]
Sure!

**Comment by jeddy3** [comment]
@evilebottnawi Would you rather we delete this PR and create a new issue for `function-url-scheme-whitelist`, or would you like to convert this PR to `function-url-scheme-whitelist`?

**Comment by evilebottnawi** [comment]
@jeddy3 convert, some busy on this week, in next week in this PR i will do this

**Comment by m-allanson** [comment]
@evilebottnawi Is it ok if I write a new PR for `function-url-scheme-whitelist`?

**Comment by evilebottnawi** [comment]
@m-allanson yep, little time now, close this

**Comment by m-allanson** [comment]
@evilebottnawi Great, thanks.

**Pull request closed without merging**